examples/training_without_local_storage.ipynb

```diff
@@ -0,0 +1,1005 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 4. Switching to Streaming Datasets\n",
+    "\n",
+    "So far, we covered how to store and load checkpoints and other artifacts with cloud storage. But, we relied on all of our training data being available at the start. While it works to download all data to a local folder before training, this technique isn't scalable for large datasets. Training wouldn't begin until _all_ data is downloaded, causing accelerators to idle.\n",
+    "\n",
+    "To overcome this bottleneck, we can switch to streaming datasets. With streaming datasets, the dataset must be split into multiple shard files, each of which contains a subset of the samples. Shards are downloaded asynchronously (in separate subprocesses) while we train. Instead of waiting until all data is downloaded, we can begin training as soon as the first shard is downloaded. This feature lets training begin instantly!\n",
+    "\n",
+    "For an in-depth walk-through on how to use streaming datasets, see our tutorial on training [FaceSynthetics with a Streaming Dataloader](https://docs.mosaicml.com/en/latest/examples/streaming_dataloader_facesynthetics.html). Here, we'll encode the MNIST dataset in the Composer streaming format and upload it to our S3 bucket."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Write the streaming dataset\n",
+    "import os\n",
+    "import torch\n",
+    "import io\n",
+    "from typing import Dict, cast, Iterable\n",
+    "import struct\n",
+    "from tqdm.auto import tqdm\n",
+    "\n",
+    "from composer.datasets.streaming import StreamingDatasetWriter\n",
+    "\n",
+    "train_out_folder = os.path.join(data_dir, 'streaming', 'train')\n",
+    "eval_out_folder = os.path.join(data_dir, 'streaming', 'eval')\n",
+    "\n",
+    "train_remote = S3ObjectStore(\n",
+    "    bucket=s3_bucket_name,\n",
+    "    prefix=bucket_prefix + '/train',\n",
+    ")\n",
+    "\n",
+    "eval_remote = S3ObjectStore(\n",
+    "    bucket=s3_bucket_name,\n",
+    "    prefix=bucket_prefix + '/eval',\n",
+    ")\n",
+    "\n",
+    "\n",
+    "shutil.rmtree(train_out_folder, ignore_errors=True)\n",
+    "shutil.rmtree(eval_out_folder, ignore_errors=True)\n",
+    "os.makedirs(train_out_folder, exist_ok=True)\n",
+    "os.makedirs(eval_out_folder, exist_ok=True)\n",
+    "\n",
+    "fields = ['i', 'x', 'y']\n",
+    "\n",
+    "def encode_sample(i: int, x: torch.Tensor, y: int) -> Dict[str, bytes]:\n",
+    "    \"\"\"Encode a (x,y) sample into a dictionary.\"\"\"\n",
+    "    x_buffer = io.BytesIO()\n",
+    "    torch.save(x, x_buffer)\n",
+    "    x_buffer.seek(0)\n",
+    "    return {\n",
+    "        'i': struct.pack('Q', i),\n",
+    "        'x': x_buffer.read(),\n",
+    "        'y': struct.pack('Q', y),\n",
```

Comment:
can we add a comment clarifying what 'Q' means and why we have "i"? I don't want users to get confused and think that adding an index is mandatory when they're probably used to just thinking about (X, y) like in sklearn, torchvision datasets, etc.

Reply:
Added comments, thanks!